generic: move bootargs patch to generic

Ansuel opened the pull request:
Different target (ipq806x, mpc85xx, mediatek, ramips) require the very same
patch. Replace these patch with more upstream friendly patch and move to
generic.

Also drop the redundant config as we can directly check for the presence
of the chosen property and the size increase is marginal.

---

~~This needs to be updated when ipq806x fixup commit gets merged as the patch needs to be removed also from there.~~

Ansuel commented on target/linux/generic/pending-5.15/921-sh-add-CONFIG_CMDLINE_OVERRIDE-option.patch:
```diff
@@ -12,14 +28,14 @@
-@@ -1162,6 +1162,17 @@ int __init early_init_dt_scan_chosen(uns
+@@ -1159,6 +1159,17 @@ int __init early_init_dt_scan_chosen(uns
  	if (p != NULL && l > 0)
- 		strlcat(data, p, min_t(int, strlen(data) + (int)l, COMMAND_LINE_SIZE));
```
This is actually different... why `strlen(data) + i` ? we are not appending here so why we are accounting for strlen of data?

neheb replied:
l is also cast to int. strlen returns size_t...

markmentovai replied:
When this patch was in target/linux/{ipq806x,mediatek,mpc85xx}/patches-5.15, [target/linux/generic/hack-5.15/920-device_tree_cmdline.patch](https://github.com/openwrt/openwrt/blob/ce014b1e3903887d47bede8128f2c698418ac5f1/target/linux/generic/hack-5.15/920-device_tree_cmdline.patch) would have been applied before it, and this context line immediately before what this patch adds would have been an append, for `bootargs-append`.

With this moving to target/linux/generic/pending-5.15, it’s applied before hack-5.15, so the patch context changes.

Patch application order is controlled by [include/kernel-build.mk](https://github.com/openwrt/openwrt/blob/53a08e37437972ba0a8fbf953a93a70a6b784ef4/include/kernel-build.mk#L12).

Ansuel commented:
@hauke some feedback about this?

DragonBluep commented:
In ramips/mt7621 target there is also a patch for `bootargs-override`😂, not sure if we can handle it here.
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/patches-5.15/314-MIPS-add-bootargs-override-property.patch
https://github.com/openwrt/openwrt/blob/master/target/linux/ramips/patches-5.10/314-MIPS-add-bootargs-override-property.patch

Ansuel commented:
@DragonBluep looks to do the same exact thing but in split code... can you test this and remove that patch? (assuming rampis works correctly on 5.15)

DragonBluep commented:
@Ansuel This is my change, will report latter.
```
 target/linux/ramips/mt7621/config-5.15        |  2 +-
 ...-MIPS-add-bootargs-override-property.patch | 63 -------------------
 2 files changed, 1 insertion(+), 64 deletions(-)
 delete mode 100644 target/linux/ramips/patches-5.15/314-MIPS-add-bootargs-override-property.patch

diff --git a/target/linux/ramips/mt7621/config-5.15 b/target/linux/ramips/mt7621/config-5.15
index 1b8a48100df..8958662d3cf 100644
--- a/target/linux/ramips/mt7621/config-5.15
+++ b/target/linux/ramips/mt7621/config-5.15
@@ -16,7 +16,7 @@ CONFIG_CLOCKSOURCE_WATCHDOG=y
 CONFIG_CLONE_BACKWARDS=y
 CONFIG_CMDLINE="rootfstype=squashfs,jffs2"
 CONFIG_CMDLINE_BOOL=y
-# CONFIG_CMDLINE_OVERRIDE is not set
+CONFIG_CMDLINE_OVERRIDE=y
 CONFIG_COMMON_CLK=y
 CONFIG_COMPAT_32BIT_TIME=y
 CONFIG_CPU_GENERIC_DUMP_TLB=y
diff --git a/target/linux/ramips/patches-5.15/314-MIPS-add-bootargs-override-property.patch b/target/linux/ramips/patches-5.15/314-MIPS-add-bootargs-override-property.patch
deleted file mode 100644
index 07635a9d86a..00000000000
........
```

DragonBluep commented:
@Ansuel Not work on mips target, we need to keep mips override patch since there is already a buggy implementation of CMDLINE_OVERRIDE.
https://github.com/torvalds/linux/blob/8bb7eca972ad531c9b149c0a51ab43a417385813/arch/mips/kernel/setup.c#L556-L559
And x86 target also has an CMDLINE_OVERRIDE, I think that's why there are several similar patches.

DragonBluep commented:
If `CONFIG_CMDLINE_OVERRIDE` is enabled in mips target, cmdline will be forcibly replaced with the `rootfstype=squashfs,jffs2`.
https://github.com/openwrt/openwrt/blob/80a62a675d2ef57a1885e1829926dd2ea1682612/target/linux/ramips/mt7621/config-5.15#L17

Ansuel commented:
Ok so we the problem here is that the is a clash of configs... probably a better name for this is CMDLINE_ALTERNATIVE or some kind of fallback???

Think i need the help of more devs to sort this out. @hauke any hint?

Problem:
- Upstream each target have the same config COMLINE_OVERRIDE that ignore any chosen value and just apply a string from the kernel config file
- Downstream we use the same config to replace the bootargs from bootargs-override in dt with the function scan_chosen...

On generalizing this the 2 config clash and we have a behaviour change.

Ansuel committed:
generic: move bootargs patch to generic

Different target (ipq806x, mpc85xx, mediatek, ramips) require the very same
patch. Replace these patch with more upstream friendly patch and move to
generic.

Also drop the redundant config as we can directly check for the presence
of the chosen property and the size increase is marginal.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

hauke commented:
I think this looks ok. I prefer to have the overwrite settings in dts too, then we can activate it on a per device basis. 

Getting this upstream could be a challenge. I ran into similar problems some years ago, there someone said that we should always use what the bootloader says because people want to control the bootargs over the bootloader. I think you should stress the special case of OpenWrt where we just replace the Linux and keep the (sometimes broken) vendor bootloader.

hauke commented:
> > there someone said that we should always use what the bootloader says because people want to control the bootargs over the bootloader
> 
> that is plain bullshit as 99% of the times bootloader are locked also this still keeps bootargs... just gives option... hope they changed mind upstream but these reason for rejecting this sounds a bit fragile to me
> 
> Very curious what they will say

I think it was not about this patchset, but an other which overwrite some settings from bootargs in some conditions. If you can change the bootloader depends from where you are coming. If you are the vendor of a product you can make the bootloader work as expected.
